package/gluon-mesh-batman-adv-core/src/respondd.c

```diff
@@ -0,0 +1,606 @@
+
+        glob_t lower;
+        if (!glob(pattern, GLOB_NOSORT, NULL, &lower)) {
+                size_t i;
+                for (i = 0; i < lower.gl_pathc; i++) {
+                        mesh_add_subif(lower.gl_pathv[i] + pattern_len - 1,
+                                       wireless, tunnel, other);
+                }
+
+                globfree(&lower);
+        }
+
+        struct json_object *ret = json_object_new_object();
+        add_if_not_empty(ret, "wireless", wireless);
+        add_if_not_empty(ret, "tunnel", tunnel);
+        add_if_not_empty(ret, "other", other);
+        return ret;
+}
+
+static struct json_object * get_mesh(void) {
+        struct json_object *ret = json_object_new_object();
+        struct json_object *bat0_interfaces = json_object_new_object();
+        json_object_object_add(bat0_interfaces, "interfaces", get_mesh_subifs("bat0"));
+        json_object_object_add(ret, "bat0", bat0_interfaces);
+        return ret;
+}
+
+static struct json_object * get_batman_adv_compat(void) {
+        FILE *f = fopen("/lib/gluon/mesh-batman-adv-core/compat", "r");
+        if (!f)
+                return NULL;
+
+        struct json_object *ret = NULL;
+
+        int compat;
+        if (fscanf(f, "%i", &compat) == 1)
+                ret = json_object_new_int(compat);
+
+        fclose(f);
+
+        return ret;
+}
+
+static struct json_object * respondd_provider_nodeinfo(void) {
+        struct json_object *ret = json_object_new_object();
+
+        struct json_object *network = json_object_new_object();
+        json_object_object_add(network, "addresses", get_addresses());
+        json_object_object_add(network, "mesh", get_mesh());
+        json_object_object_add(ret, "network", network);
+
+        struct json_object *software = json_object_new_object();
+        struct json_object *software_batman_adv = json_object_new_object();
+        json_object_object_add(software_batman_adv, "version", gluonutil_wrap_and_free_string(gluonutil_read_line("/sys/module/batman_adv/version")));
+        json_object_object_add(software_batman_adv, "compat", get_batman_adv_compat());
+        json_object_object_add(software, "batman-adv", software_batman_adv);
+        json_object_object_add(ret, "software", software);
+
+        return ret;
+}
+
+
+static void add_gateway(struct json_object *obj) {
+        FILE *f = fopen("/sys/kernel/debug/batman_adv/bat0/gateways", "r");
+        if (!f)
+                return;
+
+        char *line = NULL;
+        size_t len = 0;
+
+        while (getline(&line, &len, f) >= 0) {
+                char addr[18];
+
+                if (sscanf(line, "=> %17[0-9a-fA-F:]", addr) != 1)
+                        continue;
+
+                json_object_object_add(obj, "gateway", json_object_new_string(addr));
+                break;
+        }
+
+        free(line);
+        fclose(f);
+}
+
+static inline bool do_ioctl(int fd, struct ifreq *ifr, void *data) {
```

Comment:
I'll add an `ethtool_` prefix to a few functions. I've copied the code from our lua-ethtool module we used so far, where ethtool prefixes didn't make much sense as everything was ethtool ;)